### ISSUE TITLE
How to make the importer work

### ISSUE THREAD

**Issue by kikodev714** [skybrud/Skybrud.Umbraco.Redirects#184]
Skybrud Redirects Version 4.0.14
Skybrud Redirects Import Version 4.0.0
Umbraco Version 10.6.1

I'm trying the importer for the first time but I couldn't get it to work. I have a few questions

1. Do you have some public documentations somewhere that discusses How To's?
2. What are the columns required by the import? So far I'm using: 

![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/5c7cc2fb-25da-49fa-b86b-4cfe83c32c77)

3. This imports fine and successful..

![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/748c118b-c11c-406e-9cb3-bf212d2fc7c2)

![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/f0786a2c-e8d2-4ae3-9907-9879eb479163)

but the redirects are **not working**

![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/cc661744-4b70-4f6f-b884-ca99e43e3269)

I wonder if I'm missing something?

**Comment by abjerner** [comment]
Hi @kikodev714 

Judging from your screenshots and the fact that the redirects show up in the dashboard, I'd say the import it self is working as intended. The issue is more likely something in redirect middleware of the main redirects package and something in your setup. But if you can share how the redirects look in the `SkybrudRedirects` table in the database, I can help confirming that everything looks correct.

I notice that your redirects are created as permanent redirects, in which case they are cached pretty hard by your browser. Does clearing your browser cache or using Incognito mode change anything?

In your last screenshot, I'm also noticing the `#` at the end of the URL. Do you have a custom URL provider that fails calculating the destination URL? If the destination is either a content or media item, the redirects middleware will try to find the current URL of that item (opposed to redirecting to an old URL, if the name of the page has changed). If the underlying Umbraco logic fails to calculate the URL. IIRC the destination URL shown in the dashboard is calculated the same way, and this angle is probably a longshot.

My test site is still using `.UseStatusCodePagesWithReExecute("/friendly-404-page/")` from when I was debugging https://github.com/skybrud/Skybrud.Umbraco.Redirects/issues/181, and I'm not experiencing issues.

Regarding the documentation, I'm working a new site to better document our packages. But our project managers keep finding other work, so progress is unfortunately a bit slow. There is however a bit of documentation here on GitHub in the [`docs`](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/tree/docs) branch - eg. some information on the supported [columns](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/blob/docs/v4/import/columns.md).

**Comment by kikodev714** [comment]
Hi @abjerner,

Thank you for your quick response. Here are the contents of the [SkybrudRedirects] table as per your request

[Skybrud_DB.csv](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/files/12485214/Skybrud_DB.csv)

For this setup, I haven't used the redirects middleware from issue: https://github.com/skybrud/Skybrud.Umbraco.Redirects/issues/181
as we are still in the process of testing that (so far it looks like this one is working. :-))

I did mention though from issue 181 that we are using USkinned 3 Sitebuilder and they have internal 404 handlers using Umbraco LastChanceContentFinder class however, I don't think this is the issue as manually setting up the redirects (not import) works.

I've also tried hard clear cache and running the site in incognito mode but I got the same issue

I'm also not sure why the "#" was appended to the URL. As I mentioned, I simply installed the plugin and quickly tried it. Since this is just very early, I do suspect there's something wrong with my excel input format. Perhaps I'm missing a column to use or perhaps I need to format my from column and destination column better?

Please let me know what you think

**Comment by abjerner** [comment]
Hi again

I've tried a few different things, but can't reproduce any issues. I've tried importing your CSV file, and the two redirects work as intended when I play around in my test setup.

I can't think of why these redirects don't work when you test at your end, but I'm still inclined to think this isn't an issue with the import package, but more the setup related to the main redirects package (and possible your 404 error setup). Similar to the CSV file you've shared already, can you share one with examples on redirects that does work? Then I can use that for comparisons.

You can also try using the Razor partial below (remember to change the domain in the URI). This partial should try to replicate how the redirects middleware tries to find a matching redirect. Ideally you should see some information about the redirect.

```cshtml
@using Skybrud.Umbraco.Redirects.Services
@inject IRedirectsService RedirectsService
@{

    Uri uri = new Uri("https://localhost:44307/newlinksample.aspx");

    var redirect = RedirectsService.GetRedirectByUri(uri);

    if (redirect is null) {
        <em>No redirect found for <u>@uri</u></em>
        return;
    }

    string destinationUrl = RedirectsService.GetDestinationUrl(redirect);

    <h3>Redirect</h3>

    <table class="table details">
        <tr>
            <th>ID</th>
            <td>@redirect.Id</td>
        </tr>
        <tr>
            <th>Key</th>
            <td>@redirect.Key</td>
        </tr>
        <tr>
            <th>ID</th>
            <td>@redirect.Url</td>
        </tr>
        <tr>
            <th>Query String</th>
            <td>@redirect.QueryString</td>
        </tr>
        <tr>
            <th>Destination Type</th>
            <td>@redirect.Destination.Type</td>
        </tr>
        <tr>
            <th>Destination URL</th>
            <td>@redirect.Destination.Url</td>
        </tr>
    </table>

    <h3>Destination URL</h3>
    <pre>@destinationUrl</pre>

}
```

**Comment by kikodev714** [comment]
Hi @abjerner 

Thanks for the response. 
Adding redirects manually is the one that still works. Here's that entry:
![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/3728e6d8-8adc-4f16-af59-4a886678d28f)

DB:

![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/dd1fe545-1aaf-41e0-923f-52e2e583372a)

Okay regarding that Razor page, I think I found something interesting:
(Left is not working from import) - (Right is working from manual setup)
![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/ea31787a-53f8-44b4-912d-5086a18a4dce)

If you noticed, the destination URL from the left (not working from import) has **#**
Not sure why the import is adding that.
DB entry of that is this:
![image](https://github.com/skybrud/Skybrud.Umbraco.Redirects.Import/assets/140512863/a74d1fde-594a-4283-88cd-ba06ce2f48c9)

Which doesn't really tell me anything.

Also, if I open that entry from Skybrud and remove + re-add the destination url, it will work.

Thank you for the help.

**Comment by abjerner** [comment]
@kikodev714 sorry for the lack of reply - I've been caught up in too many other things.

I notice that the last redirect in your screenshot of the database table has `en-us` for the culture, while the others don't have any value for the culture. So it could seem that there might be an issue in either **Skybrud Redirects** or **Skybrud Redirects Import** related to culture variant content. Calling `.Url(culture: "")` on a culture variant page gives you `#`.

For now you should be able to fix this by adding a `Culture` column to the CSV, add `en-US` or similar for each culture variant page, and then doing another import.

I'll also try to find some time for fixing this 😉

**Comment by kikodev714** [comment]
@abjerner  Thank you. Adding the "Culture" column works. Please keep me posted once you have a fix for the main issue. Thanks.

**Comment by abjerner** [comment]
Hi @kikodev714 

Although your issue may be caused the import package and that you hadn't specified a culture in your CSV file, the main issue was in the main redirects package. If a redirect existed with culture variant destination, but without a destination culture specified, the user would be redirected t `#`.

Affected redirects would most likely have been created from outside of the package - eg. the import package. It's also possible that redirects created before [`v4.0.9`](https://github.com/skybrud/Skybrud.Umbraco.Redirects/releases/tag/v4.0.9) with a culture variant destination were affected.

In either case, the new [`v4.0.15`](https://github.com/skybrud/Skybrud.Umbraco.Redirects/releases/tag/v4.0.15) should fix these issues.

There may still also be a minor issue in the import package, as the `DestinationCulture` field in the database will be empty if no corresponding value is specified in the imported file. I haven't yet decided yet whether this is fine (due to the fix in [`v4.0.15`](https://github.com/skybrud/Skybrud.Umbraco.Redirects/releases/tag/v4.0.15)), or the import package should choose the first culture of the destination.
